stelaval/assignment2/BinaryTree.java

```diff
@@ -0,0 +1,176 @@
+    if (nodeSet.contains(key)) {
+      containsKey(root, key);
+    } else {
+      throw new Exception("Key is out of tree");
+    }
+  }
+
+  /**
+   * Checks if a node is present in a subtree
+   * 
+   * @param currentNode is the root of a subtree
+   * 
+   * @param key is a node which ancestors will be printed
+   */
+  private boolean containsKey(Node<T> currentNode, Node<T> key) {
+    if (currentNode == null) {
+      return false;
+    } else if (currentNode.getValue().equals(key.getValue())) {
+      return true;
+    } else if (containsKey(currentNode.getLeft(), key)
+        || containsKey(currentNode.getRight(), key)) {
+      System.out.print(currentNode.getValue() + " ");
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Finds the lowest common ancestor of two nodes in a binary tree
+   * 
+   * @param node1 is the first node
+   * @param node2 is the second node
+   * 
+   * @return node that is the lowest common ancestor of node1 and node2
+   * 
+   * @throws Exception if either one or both nodes are not in the binary tree
+   */
+  public Node<T> getLowestCA(Node<T> node1, Node<T> node2) throws Exception {
+    if (nodeSet.contains(node1) && nodeSet.contains(node2)) {
+      return getCA(root, node1, node2);
+    } else {
+      throw new Exception("Nodes are not in the tree");
+    }
+  }
+
+  private Node<T> getCA(Node<T> currentNode, Node<T> node1, Node<T> node2) {
+    if (currentNode == null) {
+      return null;
+    }
+    if (currentNode.equals(node1) || currentNode.equals(node2)) {
+      return currentNode;
+    }
+
+    Node<T> leftCA = getCA(currentNode.getLeft(), node1, node2);
+    Node<T> rightCA = getCA(currentNode.getRight(), node1, node2);
+
+    if (leftCA != null && rightCA != null) {
+      return currentNode;
+    } else if (leftCA != null) {
+      return leftCA;
+    } else {
+      return rightCA;
+    }
+  }
+
+
+  private static class Node<T> {
+
+    private T value;
+    private Node<T> left;
+    private Node<T> right;
+
+    public Node(T value) {
+      this.value = value;
+      left = null;
+      right = null;
+    }
+
+    public T getValue() {
+      return value;
+    }
+
+    public Node<T> getLeft() {
+      return left;
+    }
+
+    public void setLeft(Node left) {
+      this.left = left;
+    }
+
+    public Node<T> getRight() {
+      return right;
+    }
+
+    public void setRight(Node right) {
+      this.right = right;
+    }
+  }
+
+  public static void main(String[] args) {
+    try {
+      Node<Integer> root = new Node<>(0);
+      Node<Integer> node1 = new Node<>(1);
+      Node<Integer> node2 = new Node<>(2);
+      Node<Integer> node3 = new Node<>(3);
+      Node<Integer> node4 = new Node<>(4);
+      Node<Integer> node5 = new Node<>(5);
+      Node<Integer> node6 = new Node<>(6);
+      Node<Integer> node7 = new Node<>(5);
+      BinaryTree<Integer> bTree = new BinaryTree<>();
+      root.setLeft(node1);

```

Comment:
This manual tree construction is not really convenient to use. It would be much better if `Node` had methods (or one method) to add children with given values.
One of possible solutions:
```
class Node<T> {
  /** Adds node with given value as left child and returns newly created child Node object. */
  public Node<T> addLeftChild(T value) { ... }
  ...
}
```
This would let you build tree as follows:
```
Node<Integer> root = new Node<>(0);
Node<Integer> node1 = root.addLeftChild(1);
Node<Integer> node2 = root.addRightChild(2);
...
```
It would also let you get rid of `setLeft` and `setRight` methods which make it possible for someone to break your tree structure.

Even more convenient solution: have whole creating node and adding it to the tree logic inside `BinaryTree` class, so that all that user has to do to create node and add it to the tree is calling a single method rather than two.